fluentd/lib/generate_throttle_configs/lib/generate_throttle_configs.rb

```diff
@@ -171,15 +171,27 @@ def create_default_container_input(input_conf_file, excluded, log, options={})
 <source>
   @type tail
   @id container-input
-  @label @INGRESS
   path "#{options[:cont_logs_path] || cont_logs_path}"
   pos_file "#{options[:cont_pos_file] || cont_pos_file}"
   tag kubernetes.*
   format json_or_crio
   keep_time_key true
   read_from_head "#{options[:read_from_head] || read_from_head}"
   exclude_path #{excluded}
+  @label @CONCAT
 </source>
+<label @CONCAT>
+  <filter kubernetes.**>
+    @type concat
+    key log
+    partial_key logtag
+    partial_value P
```

Comment:
when reading from a docker json-file log, and there is no field `logtag`, what happens?  I'm assuming it just passes the record through.

Reply:
Good question.  I ran the ci-test with "-e openshift_use_crio=false ", i.e., with docker.  The tests themselves passed.  But "oc logs FLUENTDPOD" returns nothing, which is unusual.  I was suspecting too many errors/warnings might have been dumped.

This is the line where the logtag value is checked.  It looks to me it causes an exception...
https://github.com/nhosoi/origin-aggregated-logging/blob/crio-0.12/fluentd/filter_concat.rb#L134

I'm going to test this modified code.
```
--- a/fluentd/filter_concat.rb
+++ b/fluentd/filter_concat.rb
@@ -131,7 +131,7 @@ module Fluent
         end
       when :partial
         @buffer[stream_identity] << [tag, time, record]
-        unless @partial_value == record[@partial_key]
+        unless record.key?(@partial_key) || @partial_value == record[@partial_key]
           new_time, new_record = flush_buffer(stream_identity)
           time = new_time if @use_first_timestamp
           new_record.delete(@partial_key)
```

Reply:
no, the old code should be fine:
```
        unless @partial_value == record[@partial_key]
```
`record['logtag']` returns `nil` - in ruby, if you access a Hash by a non-existent key, `nil` is returned by default: http://ruby-doc.org/core-2.0.0/Hash.html#method-c-new
It is only a problem if `record` is not a Hash (and that will cause many more problems . . .)

Reply:
since `"P" == nil` will always return `false`, you will always go through the `unless` block

Reply:
> "oc logs FLUENTDPOD" returns nothing,

do you have a log file for the fluentd pod under `/var/log/containers/`?  Or, if using journald, do you have a journalctl -o export log matching `CONTAINER_NAME=.*fluentd`?

Reply:
Good to hear about the original code.  I'll revert it.
Sorry, I've already terminated the instance...  
I'm running with "LOG_DRIVER=json-file".

Reply:
Yep, I do.  And every time fluentd pod is restarted, an empty log file logging-fluentd-*.log is created like this...
```-rw-r-----. 1 root root 0 Sep 19 21:58 /var/log/containers/logging-fluentd-lpq4s_openshift-logging_fluentd-elasticsearch-136d496ae499b34e415a6e998fa9b95dda1e5171494cb007a02d60686511f96c.log```
Next, I set DEBUG=true.
```oc set env ds/logging-fluentd DEBUG=true```
but still the log is empty...
```
# oc rsh logging-fluentd-k27xq
sh-4.2# ls -l /var/log/fluentd.log 
-rw-r--r--. 1 root root 0 Sep 19 22:04 /var/log/fluentd.log
```